src/DataStructures/Tensor/Expressions/TensorAsExpression.hpp

```diff
@@ -0,0 +1,363 @@
+                                                  tmpl::pin<Lhs>>>;
+
+namespace detail {
+template <typename Element, typename Iteration, typename Lhs, typename Rhs,
+          typename RhsWithOnlyLhs, typename IndexInLhs>
+struct generate_transformation_helper {
+  using tensor_index_to_find = tmpl::at<RhsWithOnlyLhs, IndexInLhs>;
+  using index_to_replace_with = tmpl::index_of<Rhs, tensor_index_to_find>;
+  using type = tmpl::size_t<index_to_replace_with::value>;
+};
+
+template <typename Element, typename Iteration, typename Lhs, typename Rhs,
+          typename RhsWithOnlyLhs>
+struct generate_transformation_helper<Element, Iteration, Lhs, Rhs,
+                                      RhsWithOnlyLhs, tmpl::no_such_type_> {
+  using type = tmpl::size_t<Iteration::value>;
+};
+
+template <typename State, typename Element, typename Iteration, typename Lhs,
+          typename Rhs, typename RhsWithOnlyLhs>
+struct generate_transformation_impl {
+  using index_in_lhs = tmpl::index_of<Lhs, Element>;
+  using type = tmpl::push_back<State, typename generate_transformation_helper<
+                                          Element, Iteration, Lhs, Rhs,
+                                          RhsWithOnlyLhs, index_in_lhs>::type>;
+};
+}  // namespace detail
+
+/// \ingroup TensorExpressionsGroup
+/// \brief Generate transformation to account for index order difference in RHS
+/// and LHS.
+///
+/// \details
+/// Generates the transformation \f$\mathcal{T}\f$ that rearranges the Tensor
+/// index array to account for index order differences between the LHS and RHS
+/// of the tensor expression.
+///
+/// ### Usage
+/// For typelists `Rhs`, `Lhs` and `RhsOnyWithLhs`, where `RhsOnlyWithLhs` is
+/// the result of the metafunction rhs_elements_in_lhs,
+/// \code{.cpp}
+/// using result = generate_transformation<Rhs, Lhs, RhsOnlyWithLhs>;
+/// \endcode
+/// \metareturns
+/// typelist
+template <typename Rhs, typename Lhs, typename RhsOnyWithLhs>
+using generate_transformation = tmpl::enumerated_fold<
+    Rhs, tmpl::list<>,
+    detail::generate_transformation_impl<tmpl::_state, tmpl::_element, tmpl::_3,
+                                         tmpl::pin<Lhs>, tmpl::pin<Rhs>,
+                                         tmpl::pin<RhsOnyWithLhs>>>;
+
+/// \ingroup TensorExpressionsGroup
+/// \brief Defines an expression representating a Tensor
+///
+/// \details
+/// In order to represent a tensor as an expression, instead of having Tensor
+/// derive off of TensorExpression, a TensorAsExpression derives off of
+/// TensorExpression and contains a pointer to a Tensor. The reason having
+/// Tensor derive off of TensorExpression is problematic is that the index
+/// structure is part of the type of the TensorExpression, so every possible
+/// permutation and combination of indices must be derived from. For a rank 3
+/// tensor, this is already over 500 base classes, which the Intel compiler
```

Comment:
Do we want to keep this clause about the Intel compiler? We dropped support for this compiler some time ago... pinging @nilsdeppe for thoughts on this.

Reply:
I think the value in keeping it is that we know why the current design was made. I'm still optimistic that some day we can use the Intel compiler :P

Reply:
Sounds good!